presto-main/src/test/java/com/facebook/presto/operator/scalar/TestUrlFunctions.java

```diff
@@ -24,29 +24,50 @@
     @Test
     public void testUrlExtract()
     {
-        validateUrlExtract("http://example.com/path1/p.php?k1=v1&k2=v2#Ref1", "http", "example.com", null, "/path1/p.php", "k1=v1&k2=v2", "Ref1");
-        validateUrlExtract("http://example.com/path1/p.php?", "http", "example.com", null, "/path1/p.php", "", "");
-        validateUrlExtract("http://example.com/path1/p.php", "http", "example.com", null, "/path1/p.php", "", "");
-        validateUrlExtract("http://example.com:8080/path1/p.php?k1=v1&k2=v2#Ref1", "http", "example.com", 8080L, "/path1/p.php", "k1=v1&k2=v2", "Ref1");
-        validateUrlExtract("https://username@example.com", "https", "example.com", null, "", "", "");
-        validateUrlExtract("https://username:password@example.com", "https", "example.com", null, "", "", "");
-        validateUrlExtract("mailto:test@example.com", "mailto", "", null, "", "", "");
-        validateUrlExtract("foo", "", "", null, "foo", "", "");
-        validateUrlExtract("http://example.com/^", null, null, null, null, null, null);
+        validateUrlExtract("http://example.com/path1/p.php?k1=v1&k2=v2#Ref1", "http", "example.com", null, "/path1/p.php", "k1=v1&k2=v2", "Ref1", "k1=v1&k2=v2");
+        validateUrlExtract("http://example.com/path1/p.php?", "http", "example.com", null, "/path1/p.php", "", "", "");
+        validateUrlExtract("http://example.com/path1/p.php", "http", "example.com", null, "/path1/p.php", "", "", "");
+        validateUrlExtract("http://example.com:8080/path1/p.php?k1=v1&k2=v2#Ref1", "http", "example.com", 8080L, "/path1/p.php", "k1=v1&k2=v2", "Ref1", "k1=v1&k2=v2");
+        validateUrlExtract("https://username@example.com", "https", "example.com", null, "", "", "", "");
+        validateUrlExtract("https://username:password@example.com", "https", "example.com", null, "", "", "", "");
+        validateUrlExtract("mailto:test@example.com", "mailto", "", null, "", "", "", "");
+        validateUrlExtract("foo", "", "", null, "foo", "", "", "");
+        validateUrlExtract("http://example.com/^", null, null, null, null, null, null, null);
+        validateUrlExtract("http://example.com/path1/p.php?source_url=http%3a%2f%2fbar.com%3fa%3dfoo%26b%3dbar&k1=v1#Ref1", "http", "example.com", null, "/path1/p.php", "source_url=http://bar.com?a=foo&b=bar&k1=v1", "Ref1", "source_url=http%3a%2f%2fbar.com%3fa%3dfoo%26b%3dbar&k1=v1");
```

Comment:
Let's add a test for `query` and `rawQuery` that contains a `+` in the original URL. Something like this: `http://example.com/p.php?abc=hello+world`